### DISCUSSION TITLE
How to play multiple 1s notes simultaneous with aftertone (for musical application) without great memory leak

### DISCUSSION THREAD

**Discussion by iatndiag** [bluefireteam/audioplayers#1461] [Q&A]
How would I describe the problem:
Each audio player can only play the file it currently captures (As if he wants to write something down there) and prevents another audio player from capturing it. Therefore, if you want to click on the button five times per second and hear the simultaneous sound of all sounds and the after sound from the previous ones, you need five new audio players (hash table key:audioPlayer()), each of which plays its own file, each from its own folder (each will be copied to a temporary directory if the filename is the same). The audio player will not give access to the file until its sound is completed. Or you will have to do MyTableOfAudioplayers[i].dispose() breaking off every unfinished sound to stop playback and return open access to the file to other audio players. The problem is caused by the lack of access to the still-sounding file (OS error: file is busy by another process).

The result is either a memory leak of more than 1 GB and a small folder with 60 MB of audio files, or no memory leak at all (only 120 MB of RAM), but with a lot of temporary folders (for example 60 MB * 10). I repeat, for two dozen notes, after two dozen playback cycles of several dozen notes each time (with the creation of new copies of the Audio Player), the memory leak will already be more than 1 GB

And you can’t do .dispose() during playback, otherwise - a red or gray screen (a serious mistake, almost a crash). Also, if you do not .dispose(), but do .release(), then the Audio Player will no longer get access to its file, even if you try to return it to it. We need to create a new copy of the audio player, adding it to our audioPlayers hash table. This is how I understand what's going on. If there is a way to allow multiple processes to read the same file without leaking memory, then that WOULD BE the BEST solution.

**Comment by iatndiag** [comment]
I Found the solution! This is my post. No Memory Leak At All:

```dart

Map<String,AudioPlayer> audioPlayersMap1 = {}; // Map of String:audioplayers, to dispose each one personally by key

.................

  break;
  case 4:
    switch (ext) {
      case 1: _xtsn = 'WAV'; _aP = 'wavC/';  break;      case 2: _xtsn = 'm4a'; _aP = 'm4aC/';  break;
      case 3: _xtsn = '';    _aP = '';      break;      case 4: _xtsn = 'mp3'; _aP = 'mp3C/';  break;
      default: {  _xtsn = 'WAV'; _aP = 'wavC/';  } break;
    } //end switch
    _kM = krSnd[_tI][_nI][_sOl].toString();
    _nT = _aP + krSnd[_tI][_nI][_sOl] + '.' + _xtsn; // = we have got a ready path to sound file
    if(audioPlayersMap4.containsKey(_kM)) {} else {
      audioPlayersMap4.addEntries({_kM : AudioPlayer()}.entries);  // adding AudioPlayer for absolutely new note
      audioPlayersMap4 = Map.fromEntries(audioPlayersMap4.entries.toList()..sort((e1, e2) => e1.key.compareTo(e2.key))); // sorting HashMap
    } //end if
    audioPlayersMap4[_kM]?.setVolume(nVol); audioPlayersMap4[_kM]?.play(AssetSource('$_nT'), mode: PlayerMode.lowLatency);
    setState((){countAmap=5;});
    break;
  case 5:

...........

/// JUST USE STOP() !!!!!!


    audioPlayersMap[_kM]?.setVolume(nVol); await audioPlayersMap[_kM]?.stop();
     await audioPlayersMap[_kM]?.play(AssetSource('$_nT'), mode: PlayerMode.lowLatency);



```

**Reply by iatndiag** [reply]
And just one hash table is enough! All RAM consumption is only 100 MB!!! It's just great! And no memory leak!

**Reply by iatndiag** [reply]
Memory Leak was ONLY in Windows build. In Android build No memory leak At All.

In Android build simple construction works without Memory Leak At All:

```dart
    await player.play(AssetSource('$_nT'), mode: PlayerMode.lowLatency);  // Sounds From Here, Mode for each note: lowLatency ! But Memory Leak Windows!

``` `
```

**Reply by iatndiag** [reply]
In android, the previous version for Windows will not work. As soon as one and the same note is repeated, there will be no access to the file from the OS side and silence will sound. (The file is in use by another process). But this simple construction works in Android, but causes a memory leak in Windows. Therefore, the previous hash table works well in the Windows assembly.

Solution found!

**Comment by Gustl22** [comment]
We will soon release [AudioPool](https://github.com/bluefireteam/audioplayers/blob/main/packages/audioplayers/lib/src/audio_pool.dart), where you can create a set players with the same sound. Hopefully it is not mess up the memory.
